crates/subspace-service/src/dsn/import_blocks/piece_validator.rs

```diff
@@ -0,0 +1,79 @@
+use async_trait::async_trait;
+use std::collections::BTreeMap;
+use subspace_archiving::archiver::is_piece_valid;
+use subspace_core_primitives::crypto::kzg::Kzg;
+use subspace_core_primitives::{
+    Piece, PieceIndex, RecordsRoot, RootBlock, SegmentIndex, PIECES_IN_SEGMENT, RECORD_SIZE,
+};
+use subspace_networking::libp2p::PeerId;
+use subspace_networking::utils::piece_provider::PieceValidator;
+use subspace_networking::Node;
+use tracing::error;
+
+pub struct RecordsRootPieceValidator {
+    dsn_node: Node,
+    kzg: Kzg,
+    records_root_cache: BTreeMap<SegmentIndex, RecordsRoot>,
+}
+
+impl RecordsRootPieceValidator {
+    pub fn new(dsn_node: Node, kzg: Kzg, root_blocks: Vec<RootBlock>) -> Self {
+        let records_root_cache = BTreeMap::from_iter(
+            root_blocks
+                .iter()
+                .map(|rb| (rb.segment_index(), rb.records_root())),
+        );
```

Comment:
I think this is the same as:
```rust
let records_root_cache = root_blocks
                .iter()
                .map(|rb| (rb.segment_index(), rb.records_root()))
                .collect();
```

Very strange that clippy didn't complain about it.